### PR TITLE
Subprotocol header

### PR DESCRIPTION
**Pull request by conorbros**
Add client-side handling for subprotocol header during handshake.

### CODE REVIEW

**Comment by conorbros** [comment]
@dnaka91 Are you interested in merging this?

**Comment by daniel-abramov** [comment]
Hey, I think we've forgotten about this one, sorry (though the code looks familiar as if I have already skewed through it once).

Generally, it looks ok - there is an intersection with https://github.com/snapview/tungstenite-rs/pull/400 though, we'll need to check how to merge it without causing too many conflicts.

**Comment by conorbros** [comment]
Thanks for getting back to me. I'll update it once #400 is progressed.

**Comment by Its-Just-Nans** [comment]
Hello @conorbros 

Here is a PR to merge https://github.com/shotover/tungstenite-rs/pull/1 to sync with the repo

File: tests/handshake.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#![cfg(feature = "handshake")]
+use std::net::TcpListener;
+use std::thread::{sleep, spawn};
+use std::time::Duration;
+use tungstenite::error::{Error, ProtocolError, SubProtocolError};
+use tungstenite::handshake::client::generate_key;
+use tungstenite::handshake::server::{Request, Response};
+use tungstenite::{accept_hdr, connect};
```
Comment (Its-Just-Nans) [inline tests/handshake.rs:8]:
```suggestion
 #![cfg(feature = "handshake")]
-use std::net::TcpListener;
-use std::thread::{sleep, spawn};
-use std::time::Duration;
-use tungstenite::error::{Error, ProtocolError, SubProtocolError};
-use tungstenite::handshake::client::generate_key;
-use tungstenite::handshake::server::{Request, Response};
-use tungstenite::{accept_hdr, connect};
+use std::{
+    net::TcpListener,
+    thread::{sleep, spawn},
+    time::Duration,
+};
+use tungstenite::{
+    accept_hdr, connect,
+    error::{Error, ProtocolError, SubProtocolError},
+    handshake::{
+        client::generate_key,
+        server::{Request, Response},
+    },
+};
 
 fn create_http_request(uri: &str, subprotocols: Option<Vec<String>>) -> http::Request<()> {
     let uri = uri.parse::<http::Uri>().unwrap();
```

Comment (Its-Just-Nans) [inline tests/handshake.rs:8]:
```suggestion
use std::{
    net::TcpListener,
    thread::{sleep, spawn},
    time::Duration,
};
use tungstenite::{
    accept_hdr, connect,
    error::{Error, ProtocolError, SubProtocolError},
    handshake::{
        client::generate_key,
        server::{Request, Response},
    },
};
```

**Review by Its-Just-Nans — CHANGES_REQUESTED** [review]
Hi @conorbros 

Can you please juste made this tiny changes to passe the CI ?

FYI CI error is here : https://github.com/snapview/tungstenite-rs/actions/runs/8841412949/job/24741691622

File: tests/handshake.rs (hunk 1 above)
Comment (rukai) [reply tests/handshake.rs]: Since I have access to that repo, I've clicked merged for you.

Comment (Its-Just-Nans) [reply tests/handshake.rs]: Thanks !
This should fix the CI !

**Comment by agalakhov** [comment]
Ok, now it looks good. All tests are passed, the new code is covered with tests as well. Its time to merge.

**Pull request merged**